File: solution/ts/week01-code-analysis/src/storeAccount.ts
import {Movie} from "./movie";

export class StoreAccount {
    public totalSold: number = 0;
    // Implementation leak: MovieSale exposed
    public allSales: MovieSale[];

    constructor() {
        this.allSales = [];
    }

    // Consufing naming: is 'to' the customer name?
    sell(movie: Movie, to: string) {
        this.totalSold += movie.unitPrice;
        // Bug: if adding to the list fails, the total is increased regardless.
        this.allSales.push(new MovieSale(to, movie.title));
    }
}

class MovieSale {
    public clientName: string;
    public movieName: string;

    constructor(clientName: string, movieName: string) {
        this.clientName = clientName;
        this.movieName = movieName;
    }
}

File: solution/kotlin/week01/src/main/kotlin/movie/StoreAccount.kt
package movie

class StoreAccount {
    var totalSold: Double = 0.0
    // Implementation leak: MovieSale inner class exposed
    var allSales: MutableList<MovieSale> = mutableListOf()

    // Consufing naming: is 'to' the customer name?
    fun sell(movie: Movie, to: String) {
        totalSold += movie.unitPrice
        //Bug: if adding to the list fails, the total is increased regardless.
        allSales.add(MovieSale(to, movie.title))
    }

    class MovieSale(val clientName: String, val movieName: String)
}
